### ISSUE TITLE
Need TodoTest

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11005]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)| @wsanchez reported|
|-|-|
|Trac ID|trac#1450|
|Type|enhancement|
|Created|2006-01-26 05:54:39Z|

Attachments:

* [todotest.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/498/49820f6fd5c388c563f6beaef5585a2ad3dc8413/fece9142cd08cd985b966fe6c9e0c511cc9fd09e.py) (587 bytes) - added by jml on 2006-01-31 05:29:17Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1450 1450
type__enhancement enhancement
reporter__wsanchez wsanchez
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__invalid invalid
component__ 
keywords__ 
time__1138254879000000 1138254879000000
changetime__1138685357000000 1138685357000000
version__ 
owner__ 
cc__glyph cc__jml cc__wsanchez
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez commented|
|-|-|

```
#!html
<pre>
Need a TodoTest peer to SkipTest in trial.unittest so that one can implement a 
test method partially, have failures in that part of the test cause errors, and 
then raise TodoTest.  For example:

def test_foo(self):
  ...
  self.failUnless(self.isWellFormed(foo))
  # I need to parse foo and analyze it's contents, but
  # that code isn't ready yet.
  self.TodoTest("Check foo's contents and validate")
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez commented|
|-|-|

```
#!html
<pre>
I committed an addition in r15739 and added units tests in r15740.

jml- I'm told I should have run this by you first.  Please take a peek and let me 
know if this is an acceptable change.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
For completeness I agree that this is probably required, however, I think that
any tests which actually want to use it should probably be split into separate
test cases, if necessary with one test calling the other test method for set-up.

The tester has, after all, identified two distinct areas of the test, one which
may fail in some circumstances and another that may fail in others.  While it
shouldn't be required that EVERY separable circumstance be a different test
method (otherwise every line of every test would have to be its own method) it
doesn't make sense to merge lots of tests together after the effort has already
been done to separate them.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez commented|
|-|-|

```
#!html
<pre>
Certainly.  What I have in mind is tests that take a bit of setup or are 
otherwise involved to write up, but a really cursory first approximation is 
easy.  The cursory test is still valuable in that it detects that things aren't 
completely awry, but it should be improved in the future to include a more 
detailed validation.

Basically "this test is useful, but could be better."  I don't think that in all 
such cases splitting it up into two tests makes sense.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

```
#!html
<pre>
I do not think that this is a good feature for Trial's core. The todo feature is
there not to flag incomplete tests, but incomplete code. 

However, Trial _ought_ to allow you to extend TestCase in order to provide this
feature for your own tests.  This is one of the most important goals of the
Trial refactoring.

wsanchez, would you be happy to see if this is possible and simple?  If so,
assign the issue to yourself.  Otherwise, I'll have a play around when I get the
chance. 

If you don't want to have a TestCase subclass in your own code, then my best
suggestion is to just stick a '1/0' at the end of your incomplete tests, and
flag those tests as todo using the attribute.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez commented|
|-|-|

```
#!html
<pre>
Hrm.  I was thinking that incomplete code and incomplete tests are more or less 
the same thing, which is why I went down this route.  It would be useful, I think, 
to have a mechanism for marking incomplete tests, if todo is not it, but I don't 
see much value in having that as a separate feature from todo.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

```
#!html
<pre>
I have reverted the feature.
sandbox/jml/todotest.py (also attached) shows how you can implement TodoTest by
extending the base test case.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
